### PR TITLE
ENH: Improve envcorr code

### PR DESCRIPTION
**Pull request by larsoner**
Still WIP, but at least seems to be running on my machine for one subject. Changes:

1. Morph labels to the subject before extracting activations. Before it was using fsaverage labels (and hence vertices) for non-fsaverage subjects, which means that activations were extracted from the wrong areas.
2. Add Hilbert transform of the epochs, which was missing
3. Don't bother going label-degree->stc_subj->stc_fsaverage at the end -- knowing that the labels are equivalent means we can go straight from label-degree->stc_fsaverage when we want to plot later
4. Restructure to save each subject's results to HDF5 so that the script can be re-run for individual subjects by deleting their files, plotting in the end can be tweaked once it works, etc.
5. Unify subject list (include/exclude) with that of the task
6. Use mnefun precomputed empty-room inverse rather than creating it on the fly
7. Filter raw then epoch rather than filtering epochs (the latter creates many edge artifacts)
8. Use FIR rather than IIR filtering (in general [it's better](https://mne.tools/dev/auto_tutorials/preprocessing/25_background_filtering.html#filtering-basics) for offline analyses at least)
9. Use same fixed `reject` params that the task code uses (rather than autoreject)

Will update once I have nice sanity-check images for beta that look correct -- hopefully soon! Let's wait to merge until then.

### CODE REVIEW

**Comment by larsoner** [comment]
Okay the grand-average passes my beta/motor sanity check, to me this PR's result:



Looks more motor-ey than the `master` result:

**Comment by larsoner** [comment]
@ktavabi do you want to try running at your end?

**Comment by ktavabi** [comment]
Will do...

**Comment by ktavabi** [comment]
> @ktavabi do you want to try running at your end?

This branch is not running on my workstation without significant intervention. Just to confirm I am running the AEC script on data in `genz/genz_proc/resting` on DRAKE? @larsoner 

PS why did you change the repository name?

File: genz/processing/write_roi-aec_frames.py [hunk 1]
```diff
@@ -6,212 +6,172 @@
...
+df.reset_index(drop=True, inplace=True)
 df.columns = ["id", "sex"]
 df.sort_values(by="id")
 dups = df[df.duplicated("id")].index.values.tolist()
 df.drop(df.index[dups], inplace=True)
+df.id[:] = [id_.split('_', 1)[1] if isinstance(id_, str) else id_
+            for id_ in df.id]
 df.drop(df[df.id.isin(defaults.exclude)].index, inplace=True)
+df.drop(df[~df.id.isin(defaults.include)].index, inplace=True)
+assert len(df) == 108, len(df)
```
Comment (ktavabi) [inline genz/processing/write_roi-aec_frames.py:41]: `AssertionError` 106 here makes me think we're working on different files.

File: genz/defaults.py [hunk 2]
```diff
@@ -17,22 +17,15 @@
...
-    "115_9a",  # no cHPI
-    "121_9a",  # Source space missing points
-    "209_11a",  # Too few EOG events
-    "231_11a",  # twa_hp calc fail with assertion error
-    "427_15a",  # no epochs
-    "432_15a",  # Too few ECG events
-    "510_17a",  # Too few ECG events
-    "527_17a",  # Too few EOG events
-]  
+exclude = [  # these are unusable
```
Comment (ktavabi) [inline genz/defaults.py:20]: I am also hitting a handful of missing ERM files. So, I am not sure if I need to run any of the MNEFUN steps. I was under the impression @nordme has processed both resting & active datasets through epoching?

File: genz/defaults.py [hunk 3]
```diff
@@ -17,22 +17,15 @@
 archive = op.join(Path(__file__).parents[0], "archive")
 megdata = "/media/ktavabi/ALAYA/data/ilabs/genz_resting"
```
Comment (ktavabi) [inline genz/defaults.py:18]: This can't be...

File: genz/processing/write_roi-aec_frames.py [hunk 4]
```diff
@@ -6,212 +6,172 @@
...
 
-import matplotlib.pyplot as plt
-import mne
 import numpy as np
 import pandas as pd
-from autoreject import AutoReject
 from genz import defaults, funcs
-from meeg_preprocessing import config, utils
-from mne import compute_raw_covariance, read_epochs
+from meeg_preprocessing import config
```
Comment (ktavabi) [inline genz/processing/write_roi-aec_frames.py]: We don't need this

File: genz/processing/write_roi-aec_frames.py [hunk 5]
```diff
@@ -6,212 +6,172 @@
...
         hp, lp = vv
         a_lst[kk] = list()
-        # epoch raw into 5 sec trials
-        events = mne.make_fixed_length_events(raw, duration=5.0)
+        # now filter raw and re-epoch
+        hp = None if hp == 0 else hp
+        print(' Filtering ...', end='')
+        raw_use = raw.copy().filter(
+            hp, lp, l_trans_bandwidth=1, h_trans_bandwidth=1,
+            n_jobs=config.N_JOBS)
```
Comment (ktavabi) [inline genz/processing/write_roi-aec_frames.py]: n_jobs=4

File: genz/processing/write_roi-aec_frames.py [hunk 6]
```diff
@@ -6,212 +6,172 @@
...
-    print("    Loading data for %s" % subject)
-    bem_dir = os.path.join(defaults.subjects_dir, subject, "bem")
-    bem_fname = os.path.join(bem_dir, "%s-5120-bem-sol.fif" % subject)
-    src_fname = os.path.join(bem_dir, "%s-oct-6-src.fif" % subject)
+    subject = f'genz{ss}'
+    del ss
     subj_dir = os.path.join(defaults.megdata, subject)
+    inv_fname = os.path.join(
+        subj_dir, "inverse",
+        f"{subject}-meg-erm-inv.fif")
```
Comment (ktavabi) [inline genz/processing/write_roi-aec_frames.py:63]: I seem to be missing quite a few of these `erm-inv` files.

File: genz/defaults.py (hunk 3 above)
Comment (larsoner) [reply genz/defaults.py:18]: I made this directory *on my local machine* after the rsync and then symlinked to where I had stored the data after rsync.

File: genz/processing/write_roi-aec_frames.py (hunk 1 above)
Comment (larsoner) [reply genz/processing/write_roi-aec_frames.py:41]: Ahh I think I did make this change before re-running, will push a fix

**Comment by larsoner** [comment]
> Just to confirm I am running the AEC script on data in genz/genz_proc/resting on DRAKE? @larsoner...
> I am not sure if I need to run any of the MNEFUN steps
> I seem to be missing quite a few of these erm-inv files.

I didn't run on drake, I ran locally after rsyncing the necessary files. What I  suggest you do is `rsync` the files you need from drake to your machine. This is what I did, and I put them into `/media/ktavabi/ALAYA/data/ilabs/genz_resting` on my local machine. This is a trimmed version of my rsync script that is probably close to what you'd want to use -- basically you need the ERM inv, the sss_pca_fif resting raw file (so the script can epoch it), and a few of the anatomy/structural files so that the labels can be morphed:
```
#!/bin/bash -ef
rsync -a --prune-empty-dirs --partial --progress --include="genz*/" --include="*/sss_pca_fif/" --include="*/sss_pca_fif/*_rest*_raw_sss.fif" --include="*/inverse/" --include="*/inverse/*erm*-inv*.fif" --exclude="*" drake.ilabs.uw.edu:/storage/genz_active/t1/twa_hp/ /media/ktavabi/ALAYA/data/ilabs/genz_resting/
SUBJECT=genz*_*a
ROOT=subjects
rsync -a --partial --progress --include="${ROOT}" --include="${ROOT}/${SUBJECT}" --include="${ROOT}/${SUBJECT}/surf/" --include="${ROOT}/${SUBJECT}/surf/*.sphere*" --include="${ROOT}/${SUBJECT}/surf/*.white" drake:/storage/anat/${ROOT} /media/ktavabi/ALAYA/data/freesurfer/
```
You should probably `mv /media/ktavabi/ALAYA/data/ilabs/genz_resting /media/ktavabi/ALAYA/data/ilabs/genz_resting_bak` and `mv /media/ktavabi/ALAYA/data/freesurfer/genz* SOMEWHERE_ELSE` to fully remove the old files (and back them up) before rsyncing.

**Comment by larsoner** [comment]
> PS why did you change the repository name?

Do you mean https://github.com/larsoner/genz-1 as opposed to just `genz`? GitHub did this for me when I forked your repo because I already had a repo named `genz`: https://github.com/larsoner/genz

**Comment by ktavabi** [comment]
I'm trying to run this on files I pulled from DRAKE. Let me try RSYNCing with same command.

**Comment by ktavabi** [comment]
>drake.ilabs.uw.edu:/storage/genz_active/t1/twa_hp/

This is for _active_, where is the _resting_ buried? @larsoner

**Comment by larsoner** [comment]
> This is for active, where is the resting buried? @larsoner

Active and resting were unified in https://github.com/mdclarke/GenZ/pull/5, so the `sss_pca_fif` raw `_rest` files you need should be there

**Comment by ktavabi** [comment]
image

Looks like a reasonable reproduction?!

**Comment by ktavabi** [comment]
> Another potentially interesting step would be to run whatever scripts are meant to be run after this that summarize the results. I could look into trying to re-run the age-classification script that I wrote ages ago for example.

Your classification script is more or less in `mlops.py` But before that `write_pdp_html` merges ALL the covariate responses and writes out an HTML report of descriptive statistics. I'd like for us to have more of a finished product sooner rather than later, and I am not sure if the classification approach is the fastest yield. 

I will merge and start PR to...

- [ ] concatenate all sss_pca_fif raw instances from the task, 
- [ ] change the filename used to depend on this variable -- WDYM?
- [ ] code review
- [ ] reproduce descriptives with profiler
- [ ] A/B test `rest` v. `task` with Hoetellings T-test for individual subject-level difference in connectivity between "states"

If that pans out then we have results & a story, potentially across age.

**Comment by larsoner** [comment]
Sounds good. Want to look at this to see if there is some methodology we should adapt?

https://pubmed.ncbi.nlm.nih.gov/33709531/

**Comment by ktavabi** [comment]
I don't know what `z-coh` is off-hand but I think computing cross-coupling modulation index `MI` between our ROI AECs could be useful. Can you bake that in?

**Pull request merged**